TimeSeries: Improve tooltip positioning when tooltip overflows

ashharrison90 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

this PR ensures the tooltip makes maximum use of the space it has available, and we prefer a tooltip position of bottom right.

the logic is the same for both axes, but taking the y-axis as an example:
  - if the tooltip would overflow the bottom when placed below and the top when placed above, position it stickily to the end which has the least overflow
  - if the tooltip would overflow the bottom when placed below but not the top when placed above, position it above
  - if the tooltip would not overflow the bottom when placed below, then position it below

see video for example:

https://user-images.githubusercontent.com/20999846/124489937-7067e280-dda9-11eb-98e2-aa4893a62972.mov

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes https://github.com/grafana/grafana/issues/31098

**Special notes for your reviewer**:

ashharrison90 committed:
TimeSeries: Improve tooltip positioning when tooltip overflows

dprokop commented on packages/grafana-ui/src/components/VizTooltip/VizTooltipContainer.tsx:
```diff
@@ -64,23 +64,40 @@ export const VizTooltipContainer: React.FC<VizTooltipContainerProps> = ({
-        xO = tooltipMeasurement.width;
+      const overflowRight = Math.max(positionX + offsetX + tooltipMeasurement.width - width, 0);
+      const overflowLeft = Math.abs(Math.min(positionX - offsetX - tooltipMeasurement.width, 0));
+      const wouldOverflowRight = overflowRight > 0;
+      const wouldOverflowLeft = overflowLeft > 0;
+
+      const overflowBelow = Math.max(positionY + offsetY + tooltipMeasurement.height - height, 0);
+      const overflowAbove = Math.abs(Math.min(positionY - offsetY - tooltipMeasurement.height, 0));
+      const wouldOverflowBelow = overflowBelow > 0;
+      const wouldOverflowAbove = overflowAbove > 0;
+
+      if (wouldOverflowRight && wouldOverflowLeft) {
+        x = overflowRight > overflowLeft ? offsetX : width - offsetX - tooltipMeasurement.width;
+      } else if (wouldOverflowRight) {
+        x = positionX - offsetX - tooltipMeasurement.width;
+      } else {
+        x = positionX + offsetX;
       }
 
-      if (yOverflow < 0) {
-        yO = tooltipMeasurement.height;
+      if (wouldOverflowBelow && wouldOverflowAbove) {
+        y = overflowBelow > overflowAbove ? offsetY : height - offsetY - tooltipMeasurement.height;
+      } else if (wouldOverflowBelow) {
+        y = positionY - offsetY - tooltipMeasurement.height;
+      } else {
+        y = positionY + offsetY;
```
Cool that you have added this @ashharrison90 . Would be great if this could be extracted to a helper function and unit tested.

Another alternative that I think is worth trying instead of this calculation is to use react-popper with virtual reference(https://popper.js.org/react-popper/v2/virtual-elements/). We use it in few places and it was working great so far. Do you mind checking how well it works in this scenario?

dprokop reviewed (commented):
Cool that you have added this @ashharrison90 . Would be great if this could be extracted to a helper function and unit tested.

Another alternative that I think is worth trying instead of this calculation is to use react-popper with virtual reference(https://popper.js.org/react-popper/v2/virtual-elements/). We use it in few places and it was working great so far. Do you mind checking how well it works in this scenario?

ashharrison90 committed:
VizTooltip: Use react-popper, extract positioning calculation into util function + add unit tests

ashharrison90 commented:
@dprokop @mckn think i've addressed everything

switched to use `react-popper`, but still using the calculation as the inbuilt calculation doesn't seem to handle all cases as gracefully 🤔 

added unit tests as well 👍

ashharrison90 committed:
VizTooltip: Keep ref as tooltipRef

dprokop committed:
Merge branch 'main' into ash/31098

dprokop committed:
Use popper only for VizTooltip positioning

dprokop commented:
@ashharrison90 found some time today to peek at this ad managed to use react-popper only: f3250c7a1fa124eb1cb49a19ec7f820956cb7ac5. The position calculation that was a leftover felt redundant. WDYT?

mckn reviewed (commented):
I really like the code when using popper. It was really clean and neat, however when I tested it it doesn't really work in all scenarios. Have a look at the following scenario:

![position2](https://user-images.githubusercontent.com/172951/124952726-c55b6100-e014-11eb-9e4c-2d3717eb46e7.gif)

Also, I found another weird bug when playing around with the panel settings:

![not_update2](https://user-images.githubusercontent.com/172951/124952799-d310e680-e014-11eb-9789-5310fff88f1b.gif)

Should we open another issue on that one?

leeoniya commented:
> Should we open another issue on that one?

it's this one: https://github.com/grafana/grafana/issues/35266

mckn commented:
> > Should we open another issue on that one?
> 
> it's this one: #35266

Ah, ok ok - thanks 🙏🏻

ashharrison90 committed:
VizTooltip: Set altAxis to true to prevent overflow on y axis

ashharrison90 commented:
@dprokop @mckn have updated the PR to address the last little problem of the tooltip overflowing the y-axis and reduced the transition time a bit as 0.5s felt a bit sluggish. thanks for getting popper to work nicely, code is much cleaner now! 🎉

mckn reviewed (approved):
LGTM!

It is a bit flickering when I hover and move the mouse but this is way better than before! Great work on this Ash!
